compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -135,8 +135,23 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         // candidate which assumes $0 == int, one that assumes `$0 ==
         // usize`, etc. This spells an ambiguity.
 
-        let mut candidates = self.filter_impls(candidates, stack.obligation);
-
+        let candidates = self.filter_impls(candidates, stack.obligation);
+        let mut candidates = candidates
+            .into_iter()
+            .map(|c| match c {
+                ImplCandidate(impl_def) if self.tcx().is_general_impl(impl_def) => {
+                    match self.evaluate_candidate(stack, &c) {
+                        Ok(eval) if eval.may_apply() => Ok(Some(c)),
+                        Ok(_) => Ok(None),
+                        Err(OverflowError::Canonical) => Err(Overflow(OverflowError::Canonical)),
+                        Err(OverflowError::ErrorReporting) => Err(ErrorReporting),
+                        Err(OverflowError::Error(e)) => Err(Overflow(OverflowError::Error(e))),
+                    }
+                }
+                _ => Ok(Some(c)),
+            })
+            .flat_map(Result::transpose)
+            .collect::<Result<Vec<_>, _>>()?;
```

Comment:
Hmm, I see, this is a copy of the winnowing code below. What about making `if candidates.len() == 1 {` below additionally check that "this candidate *isn't* a fully-blanket `impl`", and just not `return` early?

It could be framed as "only early-succeed when the candidate is a refinement of the obligation", where e.g. "`$0: Iterator` is *not* a refinement of `$0: IntoIterator`".

Then such candidates would hit the winnowing step, which already exists and wouldn't need duplication.